apis/metal3.io/v1alpha1/baremetalhost_validation.go

```diff
@@ -201,3 +256,86 @@ func validateRootDeviceHints(rdh *RootDeviceHints) error {
 	}
 	return nil
 }
+
+// When making changes to this function for operationalstatus and errortype,
+// also make the corresponding changes in the OperationalStatus and
+// ErrorType fields in the struct definition of BareMetalHostStatus in
+// the file baremetalhost_types.go
+func checkStatusAnnotation(bmhstatus *BareMetalHostStatus) error {
+
+	operatstatus_allowed := []string{"", string(OperationalStatusOK), string(OperationalStatusDiscovered), string(OperationalStatusError), string(OperationalStatusDelayed), string(OperationalStatusDetached)}
+	errortype_allowed := []string{"", string(ProvisionedRegistrationError), string(RegistrationError), string(InspectionError), string(PreparationError), string(ProvisioningError), string(PowerManagementError)}
+
+	if !slices.Contains(operatstatus_allowed, string(bmhstatus.OperationalStatus)) {
+		return fmt.Errorf("invalid OperationalStatus='%s' in StatusAnnotation", string(bmhstatus.OperationalStatus))
+	}
+
+	if !slices.Contains(errortype_allowed, string(bmhstatus.ErrorType)) {
+		return fmt.Errorf("invalid ErrorType='%s' in StatusAnnotation", string(bmhstatus.ErrorType))
+	}
+
+	return nil
+}
+
+func validHwdDetailsAnnotation(hwdetannotaiton string, inspect string) error {
+	if hwdetannotaiton == "" {
+		return nil
+	}
+
+	if inspect != "disabled" {
+		return fmt.Errorf("inspection has to be disabled for HardwareDetailsAnnotation, check if {'inspect.metal3.io' : 'disabled'}")
+	}
+
+	objStatus := &HardwareDetails{}
+	err := json.Unmarshal([]byte(hwdetannotaiton), objStatus)
+	if err != nil {
+		return fmt.Errorf("failed to fetch Hardware Details from annotation, error=%w", err)
+	}
+
+	deco := json.NewDecoder(strings.NewReader(hwdetannotaiton))
+	deco.DisallowUnknownFields()
+	if err = deco.Decode(objStatus); err != nil {
```

Comment:
Same here: you're decoding HardwareDetails twice.

Reply:
fixed